README.md
# oocyte_maturity_segmenter

CHANGELOG.md
## 0.0.0 (September 28, 2023)
  - bump-version.sh file added
  - Merge pull request #1 from jesuslife/feature/setup_repo
  - Data and Weights folders pushed to bucket
  - data and weights added to DVC
  - Initial commit



Code/predict.py
import numpy as np
import cv2
import segmentation_models_pytorch as smp
import torch
import torchvision.transforms as tvt
import matplotlib.pyplot as plt
import argparse
import os
import sys


def main():
    # Gets the user parameters
    parser = argparse.ArgumentParser(description='script to segment the polar body')
    parser.add_argument('--input', help='path/to/input <image file>', type=str, required=True)
    parser.add_argument('--output', help='path/to/output <text file>', type=str, required=True)
    args = parser.parse_args()

    img_name = args.input
    txt_name = args.output
    base_name = os.path.basename(img_name).split('.')[0]

    out_dir = os.path.dirname(txt_name)
    if out_dir == '':
        out_dir = './'
        txt_name = os.path.join(out_dir, txt_name)

    # Sets the model weights
    det_weights_path = 'weights/oocyte_det.pt'
    seg_weights_path = 'weights/polar_body_seg.pt'

    # Checks the different src/dst paths
    if not os.path.exists(det_weights_path):
        sys.exit('Error: path to "%s" does not exist' % det_weights_path)

    if not os.path.exists(seg_weights_path):
        sys.exit('Error: path to "%s" does not exist' % seg_weights_path)

    if not os.path.exists(img_name):
        sys.exit('Error: path to "%s" does not exist' % img_name)

    if not os.path.exists(os.path.dirname(txt_name)):
        sys.exit('Error: path to "%s" does not exist' % txt_name)

    try:
        # Sets a device
        device = torch.device('cuda:0') if torch.cuda.is_available() else torch.device('cpu')

        # Creates the detection model
        print('Detecting object...')
        det_model = torch.hub.load('.',
                                   'custom',
                                   path=det_weights_path,
                                   force_reload=True,
                                   source='local')

        # Initializes the YOLOv5 model
        det_model.compute_iou = 0.6  # IoU threshold
        det_model.conf = 0.6  # Confidence threshold
        det_model.max_det = 1  # Max number of detections

        # Performs object detection with YOLOv5
        results = det_model(img_name)
        detections = results.xyxy[0].detach().cpu().numpy()

        # Checks if there is detections:
        if len(detections) > 0:
            box = detections[0][:4]
            x, y, w, h = round(box[0]), round(box[1]), round(box[2] - box[0]), round(box[3] - box[1])
        else:
            sys.exit('No oocyte detection')
        print('Done\n')

        # Creates the segmentation model
        print('Segmenting object...')

        seg_model = smp.Unet(
            encoder_name=ENCODER_NAME,
            encoder_weights='imagenet',
            in_channels=N_CHANNELS,
            classes=N_CHANNELS,
        )

        # Puts the model in a device (CPU or GPU)
        seg_model = seg_model.to(device)

        # Loads the trained model
        seg_model.load_state_dict(torch.load(seg_weights_path))  # Load trained model

        # Sets the model to evaluation mode
        seg_model.eval()
        print('Done')

        # Sets the graph parameters
        color = np.array([255, 0, 0], dtype='uint8')
        alpha = 0.4
        title = 'Model: %s  Encoder: %s \nInput image format: %s' % (MODEL_NAME, ENCODER_NAME, IMG_FORMAT)

        # Reads the input image
        color_img = cv2.imread(img_name)

        # Creates a CLAHE object for image contrast enhancing
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))

        # Applies the CLAHE algorithm to a grayscale image
        img = cv2.cvtColor(color_img, cv2.COLOR_BGR2GRAY)
        img = clahe.apply(img).astype(np.float32)

        # Crops the predicted mask
        crop_img = img[y:y + h, x:x + w]
        height_orig = crop_img.shape[0]
        width_orig = crop_img.shape[1]

        # Composes the image transformations
        transform_img = tvt.Compose([tvt.ToPILImage(),
                                     tvt.Resize((IMG_HEIGHT, IMG_WIDTH), tvt.InterpolationMode.BILINEAR),
                                     tvt.PILToTensor()])

        # Applies the transformations to the cropped image
        crop_img = transform_img(crop_img)

        # Puts the transformed image in a device
        crop_img = torch.autograd.Variable(crop_img, requires_grad=False).to(device).unsqueeze(0)

        # Performs object segmentation
        with torch.no_grad():
            prd = seg_model(crop_img.float())

        # Resizes the predicted mask
        prd = tvt.Resize((height_orig, width_orig), tvt.InterpolationMode.NEAREST)(prd[0])

        # Puts the predicted mask in a numpy array
        seg = prd.data.cpu().detach().numpy().squeeze(0)

        # Converts the mask values to integers
        seg = (seg > 0.5).astype(np.uint8)

        # Evaluates the oocyte maturity
        pb = np.count_nonzero(seg)
        if pb > MATURITY_THR:
            is_mature = '1'
        else:
            is_mature = '0'

        # Saves the results (0=Non-maturity, 1=Maturity) to a text file
        output = open(txt_name, "w")
        output.write(is_mature)
        output.close()

        color_img = cv2.cvtColor(color_img, cv2.COLOR_BGR2RGB)
        masked_img = np.where(seg[..., None], color, color_img[y:y + h, x:x + w])

        # Creates the superimposed image
        out = cv2.addWeighted(color_img[y:y + h, x:x + w], 1 - alpha, masked_img, alpha, 0)

        # Plots the crop/mask images
        fig, axes = plt.subplots(nrows=1, ncols=2, figsize=[10, 8])
        fig.suptitle(title)
        axes[0].axis('off')
        axes[0].set_title('Crop')
        axes[0].imshow(color_img[y:y + h, x:x + w])
        axes[1].axis('off')
        axes[1].set_title('Crop + predicted mask')
        axes[1].imshow(out)
        fig.tight_layout()
        plt.savefig(os.path.join(out_dir, base_name + '_result.jpg'))
        plt.close('all')

    except Exception as e:
        print('Exception: %s' % str(e))


if __name__ == '__main__':
    # Main parameters
    MODEL_NAME = 'Unet'
    ENCODER_NAME = 'resnet101'
    IMG_FORMAT = 'grayscale'
    N_CHANNELS = 1
    N_CLASSES = 1
    IMG_WIDTH = 256
    IMG_HEIGHT = 256
    MATURITY_THR = 10

    # Call to main function
    main()


Code/custom_dataset.py
import numpy as np
import cv2
from torch.utils.data import Dataset
from torch import get_rng_state, set_rng_state
from torchvision import transforms as tvt


class CustomDataset(Dataset):
    """
    This class is used to pre-process oocytes images and masks
    """

    def __init__(self, x, y, n_classes=1,
                 transform_img=tvt.Compose([tvt.ToPILImage(),
                                            tvt.Resize((256, 256), tvt.InterpolationMode.BILINEAR),
                                            tvt.PILToTensor()]),
                 transform_mask=tvt.Compose([tvt.ToPILImage(),
                                             tvt.Resize((256, 256), tvt.InterpolationMode.NEAREST),
                                             tvt.PILToTensor()])):
        self.x = x
        self.y = y
        self.n_classes = n_classes
        self.transform_img = transform_img
        self.transform_mask = transform_mask
        self.clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))

    def __len__(self):
        return len(self.x)

    def __getitem__(self, ix):
        img = cv2.imread(self.x[ix], cv2.IMREAD_GRAYSCALE)
        img = self.clahe.apply(img).astype(np.float32)
        state = get_rng_state()
        img = self.transform_img(img)

        mask = np.load(self.y[ix])
        # mask = (np.arange(self.n_classes) == mask[..., None]).astype(np.float32)
        set_rng_state(state)
        mask = self.transform_mask(mask)

        return img, mask


Code/train_segmenter.py
import numpy as np
from custom_dataset import CustomDataset
import torch
from torch.utils.data import DataLoader
from torchvision import transforms as tvt
import segmentation_models_pytorch as smp
import pandas as pd
import matplotlib.pyplot as plt
from glob import glob
import os
import sys
from tqdm import tqdm
from time import sleep


def compute_iou(outputs, labels):
    """
    Function to compute the iou score
    Args:
        outputs (tensor): predicted masks
        labels (tensor): ground-truth masks

    Returns:
        IoU score (float)
    """

    # True if value > 0.5, False otherwise
    outputs, labels = torch.sigmoid(outputs) > 0.5, labels > 0.5
    smooth = 1e-6

    # batch x num_classes x height x weight
    b, n, h, w = outputs.shape

    # Iterates over the tensor channels
    iou_list = []
    for i in range(0, n):
        # Extracts a mask per channel
        _out, _labs = outputs[:, i, :, :], labels[:, i, :, :]

        # Computes the intersection over axis 1 and 2
        intersection = (_out & _labs).float().sum((1, 2))

        # Computes the union over axis 1 and 2
        union = (_out | _labs).float().sum((1, 2))

        # Computes the IoU score and inserts the value into a list
        iou = (intersection + smooth) / (union + smooth)
        iou_list.append(iou.mean().item())

    return np.mean(iou_list)


def fit(model, device, dataloader, epochs=100, lr=1e-3):
    """
    Function to train the model and save the best and the last weights under a selected metric
    Args:
        model (torch model): NN model
        device (torch device): device (CPU or GPU)
        dataloader (torch dataloader): loads images and labels
        epochs (int): number of epochs for training
        lr (float): learning rate

    Returns:
        a dictionary with the training history
    """

    # Loss functions
    criterion1 = smp.losses.DiceLoss('binary', from_logits=True)
    criterion2 = smp.losses.SoftBCEWithLogitsLoss()

    # Optimizer
    optimizer = torch.optim.AdamW(params=model.parameters(), lr=lr, weight_decay=1e-4, amsgrad=True)

    # Learning rate scheduler
    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
        optimizer,
        mode='min',
        factor=0.5,
        patience=3,
        verbose=True,
        min_lr=1e-6
    )

    # Puts the model in a device (CPU or GPU)
    model.to(device)

    # Training history dictionary
    hist = {'loss': [], 'iou': [], 'val_loss': [], 'val_iou': []}

    # Initial validation loss reference
    val_loss_ref = 1.

    # Training loop
    for epoch in range(1, epochs + 1):
        # Loads images and labels
        bar = tqdm(dataloader['train'])

        # Initializes the loss and IoU values for training
        train_loss, train_iou = [], []

        # Sets the model to train mode
        model.train()

        # Training loop
        for images, masks in bar:
            # Puts the images and the labels in the device
            images, masks = images.to(device), masks.to(device)

            # Sets the gradients to None
            optimizer.zero_grad()

            # Model predictions
            y_hat = model(images)

            # Computes the training loss
            loss = criterion1(y_hat, masks.float()) + criterion2(y_hat, masks.float())
            # loss = criterion1(y_hat, masks.float())

            # Performs the backpropagation
            loss.backward()

            # Performs the parameter update
            optimizer.step()

            # Computes the training parameters for the current batch
            iou = compute_iou(y_hat, masks)
            train_loss.append(loss.item())
            train_iou.append(iou)
            bar.set_description("loss=%.5f iou=%.5f" % (float(np.mean(train_loss)), float(np.mean(train_iou))))
            sleep(0.1)

        # Stores the training parameters for the current epoch
        hist['loss'].append(np.mean(train_loss))
        hist['iou'].append(np.mean(train_iou))

        # Loads images and labels for validation
        bar = tqdm(dataloader['val'])

        # Initializes the loss and IoU values for validation
        val_loss, val_iou = [], []

        # Sets the model to evaluation mode
        model.eval()

        # Validation loop
        with torch.no_grad():
            for images, masks in bar:
                # Puts the images and the labels in the device
                images, masks = images.to(device), masks.to(device)

                # Model predictions
                y_hat = model(images)

                # Computes the validation loss
                loss = criterion1(y_hat, masks.float()) + criterion2(y_hat, masks.float())
                # loss = criterion1(y_hat, masks.float())

                # Stores the validation parameters for the current batch
                iou = compute_iou(y_hat, masks)
                val_loss.append(loss.item())
                val_iou.append(iou)
                bar.set_description("val_loss=%.5f val_iou=%.5f" % (float(np.mean(val_loss)), float(np.mean(val_iou))))
                sleep(0.1)

        # Stores the validation parameters for the current epoch
        hist['val_loss'].append(np.mean(val_loss))
        hist['val_iou'].append(np.mean(val_iou))

        # Prints the results at ending the epoch
        print('Epoch %d/%d loss=%.5f iou=%.5f val_loss=%.5f val_iou=%.5f' %
              (epoch,
               epochs,
               float(np.mean(train_loss)),
               float(np.mean(train_iou)),
               float(np.mean(val_loss)),
               float(np.mean(val_iou))))

        # Saves the best model
        if np.mean(val_loss) < val_loss_ref:
            print('Saving model...')
            torch.save(model.state_dict(), os.path.join(RESULTS_DIR, BEST_WEIGHT_FILE))
            val_loss_ref = np.mean(val_loss)

        # Applies the learning rate scheduler policy
        scheduler.step(np.mean(val_loss))  # Apply scheduler
        print()
        sleep(0.1)

    # Saves the last model
    print('Saving model...')
    torch.save(model.state_dict(), os.path.join(RESULTS_DIR, LAST_WEIGHT_FILE))

    return hist


def main():
    """
    Classes:
        0- Background
        1- Polar body
    """

    # Creates a directory for the current experiment (if the directory exists, the experiment is aborted)
    try:
        if not os.path.exists(RESULTS_DIR):
            os.makedirs(RESULTS_DIR)
        else:
            sys.exit('Path to %s already exists. Please, assign another directory for the experiment results' %
                     RESULTS_DIR)
    except Exception as e:
        sys.exit('Exception: %s' % str(e))

    # Composes the image transformations
    transform_img = tvt.Compose([tvt.ToPILImage(),
                                 tvt.Resize((HEIGHT, WIDTH), tvt.InterpolationMode.BILINEAR),
                                 tvt.RandomAffine(degrees=(0, 15),
                                                  translate=(0.0, 0.1),
                                                  interpolation=tvt.InterpolationMode.BILINEAR),
                                 tvt.RandomHorizontalFlip(p=0.2),
                                 tvt.RandomVerticalFlip(p=0.2),
                                 tvt.PILToTensor()])

    # Composes the mask transformations
    transform_mask = tvt.Compose([tvt.ToPILImage(),
                                  tvt.Resize((HEIGHT, WIDTH), tvt.InterpolationMode.NEAREST),
                                  tvt.RandomAffine(degrees=(0, 15),
                                                   translate=(0.0, 0.1),
                                                   interpolation=tvt.InterpolationMode.NEAREST),
                                  tvt.RandomHorizontalFlip(p=0.2),
                                  tvt.RandomVerticalFlip(p=0.2),
                                  tvt.PILToTensor()])

    # Gets the image/mask names
    img_names = glob(os.path.join(DATA_DIR, IMG_DIR, '*.jpg'))
    mask_names = glob(os.path.join(DATA_DIR, MASK_DIR, '*.npy'))

    # Gets the number of images and masks
    n_img = len(img_names)
    n_mask = len(mask_names)

    # Compares the images/masks sizes
    if n_img != n_mask:
        sys.exit('Error: inputs and targets have different sizes')

    # Sets the train data size
    train_size = round(n_img * 0.8)

    # Gets the train/validation datasets
    dataset = {
        'train': CustomDataset(img_names[:train_size],
                               mask_names[:train_size],
                               N_CLASSES,
                               transform_img,
                               transform_mask),
        'val': CustomDataset(img_names[train_size:],
                             mask_names[train_size:],
                             N_CLASSES,
                             transform_img,
                             transform_mask)
    }

    # Configures the train/validation dataloaders
    dataloader = {
        'train': DataLoader(dataset['train'], batch_size=BATCH_SIZE, shuffle=True, pin_memory=True),
        'val': DataLoader(dataset['val'], batch_size=BATCH_SIZE, shuffle=False, pin_memory=True)
    }

    # Sets the device (CPU or GPU)
    device = 'cuda:0' if torch.cuda.is_available() else 'cpu'

    # Creates the NN model
    model = smp.Unet(
        encoder_name=ENCODER_NAME,
        encoder_weights='imagenet',
        in_channels=N_CHANNELS,
        classes=N_CLASSES,
        # activation='sigmoid'
    )

    # Trains the model
    try:
        hist = fit(model, device, dataloader, epochs=N_EPOCHS, lr=LR)
        df = pd.DataFrame(hist)
        df.to_csv(os.path.join(RESULTS_DIR, CSV_FILE), sep=',', index=False, encoding='latin1')
        title = 'Model: %s  Encoder: %s  Loss function: %s\n Input image format: %s' % \
                (MODEL_NAME, ENCODER_NAME, LOSS_FUNCTION_NAME, IMG_FORMAT)
        plt.tight_layout()
        df.plot(grid=True, title=title, xlabel='Epoch', ylabel='Loss / IoU score', figsize=(10, 8))
        # plt.show()
        plt.savefig(os.path.join(RESULTS_DIR, IMG_FILE))
    except Exception as e:
        sys.exit('Exception: %s' % str(e))


if __name__ == '__main__':
    # Main parameters
    MODEL_NAME = 'Unet'
    ENCODER_NAME = 'resnet101'
    LOSS_FUNCTION_NAME = 'Dice + SoftBCEWithLogits'
    IMG_FORMAT = 'grayscale'
    DATA_DIR = 'data'
    IMG_DIR = 'crop_img'
    MASK_DIR = 'crop_mask'
    WEIGHT_DIR = 'weights'
    BEST_WEIGHT_FILE = 'best.pt'
    LAST_WEIGHT_FILE = 'last.pt'
    RUN_DIR = 'runs'
    TRAIN_DIR = 'train'
    TEST_DIR = 'test'
    CSV_FILE = 'train.csv'
    IMG_FILE = 'train.jpg'
    WIDTH = 256  # image width
    HEIGHT = 256  # image height
    LR = 1e-3
    BATCH_SIZE = 10
    N_EPOCHS = 60
    N_CLASSES = 1
    N_CHANNELS = 1
    ANGLE = 15
    N_EXP = 1
    RESULTS_DIR = os.path.join(RUN_DIR, TRAIN_DIR, str(N_EXP))
    # SEED = 42
    # np.random.seed(seed)
    # torch.manual_seed(seed)
    # torch.cuda.manual_seed(seed)

    # Main function call
    main()
